### PR TITLE
Lock revisions before adding more

### PR DESCRIPTION
**Pull request by johnswanson**
Our revision system has a race condition in the case of two concurrent requests to modify the same object. You can reproduce the exact issue here by opening two `psql` shells. In each one, run

```
begin;
insert into revision (model, model_id, user_id, timestamp, object, most_recent, metabase_version) values ('Dashboard', 1, 1, '2024-01-01T00:00:00Z', '{}', true, 'v1.0.0') returning id;
update revision set most_recent = false where model='Dashboard' and model_id = 10 and id != [[the id from above]];
```

after running all of the above, `commit` the transaction and note that **two** revisions are now marked as `most_recent`.

This is the way our revision system works: `t2/before-insert` sets `most_recent=true`, while `t2/after-insert` updates all the *other* revisions with `most_recent=false`.

The issue here is that when each transaction is running, it doesn't see the uncommitted result of the other transaction (the new revision), so the UPDATE does not set `most_recent` to `false` for either new revision.

The solution here is to just select all existing `revision`s for this model `FOR UPDATE`. This effectively locks the revisions for a single entity, so we can't concurrently update revisions for the same entity. Note that this only works if we already have at least one revision for the entity! But we create a revision when the entity is created, and it won't be possible to do *that* concurrently, so I think it's ok.

### CODE REVIEW

File: src/metabase/models/revision.clj [hunk 1]
```diff
@@ -83,7 +83,13 @@
   {:object mi/transform-json})
 
 (t2/define-before-insert :model/Revision
-  [revision]
+  [{:keys [model model_id] :as revision}]
+  (t2/query {:select [:id]
```
Comment (noahmoss) [inline src/metabase/models/revision.clj:87]: Probably worth adding a comment explaining what this query is doing?

**Comment by noahmoss** [comment]
Your fix makes sense to me in isolation — but raises two questions for me:
* Is there a way to rework the surrounding logic to avoid this issue? What if we get rid of `t2/define-after-insert`, and instead in `t2/define-before-insert` we just set `:most_recent false` for all existing rows for the model before we insert the new row? That would also create a predicate lock on the existing rows but avoid having this logic be spread across two different methods.
* Why do we even need the `most_recent` flag instead of just going off of insertion timestamps or PK ordering? (I assume there's a reason that I'm not aware of, but maybe not)

**Review by noahmoss — DISMISSED** [review]

**Comment by johnswanson** [comment]
> Is there a way to rework the surrounding logic to avoid this issue? What if we get rid of t2/define-after-insert, and instead in t2/define-before-insert we just set :most_recent false for all existing rows for the model before we insert the new row? That would also create a predicate lock on the existing rows but avoid having this logic be spread across two different methods.

I don't think our transaction isolation level (I think `READ COMMITTED`?) is sufficient to make this work. Say we have one revision A, and are concurrently creating revisions B and C.

- Revision B's transaction sees only revision A, sets it to `most_recent: false`, inserts a new row with `most_recent: true`

- Revision C's transaction sees only revision A, sets it to `most_recent: false`, inserts a new row with `most_recent: true`

I don't think we obtain a lock on revision A in either case? And testing it out, it does look like we end up with two `most_recent` revisions.

> Why do we even need the most_recent flag instead of just going off of insertion timestamps or PK ordering? (I assume there's a reason that I'm not aware of, but maybe not)

Good question. I think that would make more sense. Right now, we're doing a `JOIN` to `revision` in the collection items endpoint on `most_recent=true`. I think we'd need to instead `JOIN` to a subquery to make this work? But given our postmortem earlier, I'm kind of tempted to fix this with a smaller diff for now, and then maybe we can look at removing `most_recent` entirely in a followup?

**Review by noahmoss — APPROVED** [review]
Ok, makes sense 👍

**Pull request merged**